### DISCUSSION TITLE
Migrations:up does not exit with success

### DISCUSSION THREAD

**Discussion by jdmevo123** [mikro-orm/mikro-orm#4482] [General]
Running migrations up does not exit with success, so the command does not end.

npx mikro-orm migrations:up results in a success print of migrates up to x, but then there is no successful exit.

**To Reproduce**
1. npx mikro-orm migrations:up
2. Upon completion you need to manually terminate

**Expected behavior**
Once migration:up has executed, the command should return a success exit code.

**Comment by jdmevo123** [comment]
ive tried doing a command like this in docker to force exit, but because mikro-orm does not successfully exit, it never exits the container.

"command": [
            "/bin/bash",
            "-c",
            "npx mikro-orm migration:up && echo 'pass' || echo 'fail'"
        ],

**Reply by B4nan** [reply]
Maybe it's about what you have in the existing migration files?

**Comment by B4nan** [comment]
I saw your message on slack, but without a complete repro, it's hard to believe. Nothing changed in this field for years, the CLI never had to explicitly exit the process, and you are the only one ever complaining about such a problem, so it feels more like some other problem specific to your app. It does not behave like this on my end, at least not in a regular shell, the command finishes just fine. 

Also please don't ignore the issue template when creating bug reports, I have no idea what version/driver are you on. 

Tried with v6 and sqlite in https://github.com/mikro-orm/guide:

<img width="505" alt="image" src="https://github.com/mikro-orm/mikro-orm/assets/615580/882ec970-588a-491d-b566-efe498584dc2">

as well as v5 and mysql in https://github.com/mikro-orm/nestjs-realworld-example-app:

<img width="636" alt="image" src="https://github.com/mikro-orm/mikro-orm/assets/615580/e5887b8f-f53b-4289-b548-785f38966360">

Also tried with an actual migration being executed, also works fine:

<img width="755" alt="image" src="https://github.com/mikro-orm/mikro-orm/assets/615580/de0ffca4-5910-4dda-b7c4-6dd94a1c5515">

**Reply by jdmevo123** [reply]
I am using V5 with postgresql. 

Mine interacts the same way, except it never completes once its printed "[32mSuccessfully migrated up to the latest version[39m" it just sits there unless I force termination of the script

I run "npx mikro-orm migration:up; echo "pass" and the output i get to is:

![Screenshot 2023-06-25 162643](https://github.com/mikro-orm/mikro-orm/assets/6053506/be8b0ec9-9466-43f3-a15d-d3410b183a62)

As you can see it just gets stuck and does not exit as success to process the next command. I have to physically terminate the batch job.

**Reply by B4nan** [reply]
> I run "npx mikro-orm migration:up; echo "pass" and the output i get to is:

I think it should be `&&` instead of `;`, as in your opening post. But it looks like that behaves the same locally, so probably not a problem either.

What version are you on? That log with `== Migration...` sounds like something outdated, it looks differently in latest versions.

```
➜  nestjs-realworld-example-app git:(master) ✗ npx mikro-orm migration:up; echo 'foo'
Processing 'Migration20211219155639'
Transaction was implicitly committed, do not mix transactions and DDL with MySQL (#805)
Applied 'Migration20211219155639'
Successfully migrated up to the latest version
foo
```

Will try with postgres now.

edit: works fine with postgres too

**Reply by jdmevo123** [reply]
It says I’m on v5.2.7.

could it be an issue with postgresql?

I used ; because powershell doesn’t recognise &&.

**Reply by jdmevo123** [reply]
I upgraded to 5.7.12 and still the same issue

**Reply by jdmevo123** [reply]
if i make the migration fail, it executes as expected.
![Screenshot 2023-06-25 191654](https://github.com/mikro-orm/mikro-orm/assets/6053506/932e180c-4efa-464c-bacd-293438520a32)

**Reply by B4nan** [reply]
How is powershell relevant if you run things inside docker?

Anyway, I need to see a failing repro for this, it might be about the migration files and what is inside them, or your ORM config. 

We don't call `orm.close` explicitly from inside the CLI commands, but since it works fine on my end (and you are the first one reporting this), I'd guess that is also not the problem here. It could help, and I am not against doing it, but I want to first see a failing repro so I know what I am fixing.

**Reply by jdmevo123** [reply]
here is my ORM config:

`import { FileCacheAdapter, LoadStrategy, Options } from "@mikro-orm/core";
import { SqlHighlighter } from "@mikro-orm/sql-highlighter";
import { TsMorphMetadataProvider } from "@mikro-orm/reflection";
import { RedisCacheAdapter } from "mikro-orm-cache-adapter-redis";
import * as dotenv from "dotenv";
import path from "path";

dotenv.config({
  path: `.env`,
});

const isDevelopment: boolean =
  !process.env["NODE_ENV"] || process.env["NODE_ENV"] == "dev";

const config = {
  type: "postgresql",
  host: process.env["DATABASE_HOST"],
  port: Number(process.env["DATABASE_PORT"]),
  user: process.env["DATABASE_USER"],
  password: process.env["DATABASE_PASSWORD"],
  dbName: process.env["DATABASE_NAME"],
  entities: ["dist/**/*.entity.js"],
  entitiesTs: ["src/**/*.entity.ts"],
  debug: isDevelopment,
  highlighter: new SqlHighlighter(),
  metadataProvider: TsMorphMetadataProvider,
  migrations: {
    tableName: "migrations_data", // migrations table name
    path: path.resolve(__dirname + "/migrations"), // path to folder with migration files
    pattern: /^[\w-]+\d+\.ts$/, // how to match migration files
    transactional: true, // run each migration inside transaction
    disableForeignKeys: true, // try to disable foreign_key_checks (or equivalent)
    allOrNothing: true, // run all migrations in current batch in master transaction
    emit: "ts", // migration generation mode
  },
  cache: {
    enabled: true,
    pretty: isDevelopment, // allows to pretty print the JSON cache TODO set to false on prod
    adapter: FileCacheAdapter, // you can provide your own implementation here, e.g. with redis
    options: { cacheDir: process.cwd() + "/temp" }, // options will be passed to the constructor of `adapter` class
  },
  resultCache: {
    expiration: 2000, // 2secs by default
    adapter: RedisCacheAdapter,
    options: {
      host: process.env["REDIS_SERVER_HOST"],
      port: process.env["REDIS_SERVER_PORT"],
      db: process.env["REDIS_SERVER_DATABASE"],
      password: process.env["REDIS_SERVER_PASSWORD"],
    },
  },
  loadStrategy: LoadStrategy.JOINED,
} as Options;

export default config;
`

**Comment by Lukica00** [comment]
I am having the same problem with all cli commands. Did you find a way to solve it or cause of problem at least?

**Reply by papalardo** [reply]
Im having the same problem. I noticed it begin when I implemented resultCache with redis.
Removing resultCache its works normally.
Then, I did like this:

```ts
const isMikroCli = process.argv.some((arg) => arg.includes('@mikro-orm/cli'));

{
   resultCache: isMikroCli ? undefined : { } as RedisOptions
}
```

**Reply by B4nan** [reply]
That sounds like you never close your redis connection. Do you implement the `CacheAdapter.close` method?

**Reply by papalardo** [reply]
@B4nan yes. Im using [this package](https://github.com/ramiel/mikro-orm-cache-adapter-redis)

**Reply by B4nan** [reply]
And when you put some logs to that method, is it executed? Looking at the code it should, we call `orm.close` from the migration commands: https://github.com/mikro-orm/mikro-orm/blob/cecea34c866dcb5a567a226e011b37015d7b8052/packages/cli/src/commands/MigrationCommandFactory.ts#L120

And that method is calling the result cache `close` method:

https://github.com/mikro-orm/mikro-orm/blob/194f511ed2102db93ab5a437348097b2ebbd8bfd/packages/core/src/MikroORM.ts#L196
